### PR TITLE
Feature/#1 setup initial backend modules

### LINKED ISSUE #1: Setup initial backend modules
```
Server Side
├── Session
│   ├── id (String)
│   ├── host (User)
│   ├── Genre (Genre)
│   ├── participants (ArrayList<User>)
│   ├── ChatStream (ChatStream)
│   └── VideoInfo (VideoInfo)
│
├── PublicSession (Extends Session)
│   └── n/a
│
├── PrivateSession (Extends Session)
│   ├── password (byte[])
│   ├── max_participants (int)
│   └── vacant (boolean)
│
├── User (Class)
│   ├── id (String)
│   ├── username (String)
│   ├── avatar_url (String)
│   ├── session_hosted (int)
│   ├── messages_sent (int)
│   └── karma (int)
│
├── Genre (Enum)
│   └── // Gaming, Cooking, Coding...
│
├── ChatStream (Class)
│   ├── messages (HashMap<UserMessage.id, pair (tupe[video_timestamp (long), UserMessage])>)
│   └── participated (ArrayList<User>)
│
├── VideoInfo (Class)
│   ├── url (String)
│   ├── title (String)
│   ├── timestamp (long)
│   └── playing (boolean)
│
├── Message (Class)
│   ├── id (String)
│   ├── timestamp (long)
│   └── content (String)
│
├── UserMessage(Extends Message)
│   ├── sender (User)
│   └── flagged (boolean)
│
└── FullUser (Extends User)
    ├── token (String)
    ├── email (String)
    ├── password (byte[])
    ├── creation_date (long)
    ├── last_login (long)
    └── avatar_url (String)
```

### PR DESCRIPTION
**Pull request by Xu-Gavin**
Since we transitioned away from rust, I setup the SpringBoot framework with Gradle. I added all data models (initial backend modules) and tests accordingly. I've also added a logging plugin to help with seeing tests visually.

Closes #1.

### CODE REVIEW

File: src/main/java/com/syncstream_dev/syncstream/model/VideoInfo.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
+package com.syncstream_dev.syncstream.model;
+
+public class VideoInfo {
+    private String url;
+    private String title;
+    private Long timestamp;
+
+    public VideoInfo(String url, String title, Long timestamp)  {
```
Comment (itsmarsss) [inline src/main/java/com/syncstream_dev/syncstream/model/VideoInfo.java]: Setters may be beneficial for future development. (e.g. User switches video, all of `url`, `title` and `timestamp` will have to change)

File: src/main/java/com/syncstream_dev/syncstream/model/message/Message.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
+package com.syncstream_dev.syncstream.model.message;
+
+public abstract class Message {
+    protected String mid;     // Message id
```
Comment (itsmarsss) [inline src/main/java/com/syncstream_dev/syncstream/model/message/Message.java]: Avoid abbreviations like `mid`.

`messageId` works fine.

File: src/main/java/com/syncstream_dev/syncstream/model/session/PrivateSession.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
+package com.syncstream_dev.syncstream.model.session;
+
+import com.syncstream_dev.syncstream.model.ChatStream;
+import com.syncstream_dev.syncstream.model.Genre;
+import com.syncstream_dev.syncstream.model.VideoInfo;
+
+public class PrivateSession extends Session {
+    private String password;
+    private int maxViewers;
```
Comment (itsmarsss) [inline src/main/java/com/syncstream_dev/syncstream/model/session/PrivateSession.java:9]: If we are sticking to non-primitive types, then use `Integer`.

File: src/main/java/com/syncstream_dev/syncstream/model/user/User.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        this.userInfo = new UserInfo(token, email, password, LocalDateTime.now(), LocalDateTime.now(), avatarUrl);
+        this.uid = uid;
+        this.username = username;
+        this.messagesSent = 0;
+        this.points = 0;
+    }
+
+    // Login User Constructor
+    public User(String username, String password) {
+        throw new UnsupportedOperationException("Not implemented yet.");    // Todo: Requires database integration to load user info
```
Comment (itsmarsss) [inline src/main/java/com/syncstream_dev/syncstream/model/user/User.java]: `// Todo:` on its own line so it does not look like documentation.

P.S. `TODO` in all caps for readability.

File: src/main/java/com/syncstream_dev/syncstream/model/message/UserMessage.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+        super(mid, timestamp, content);
+        this.sender = sender;
+        this.flag = flag;
+    }
+
+    public User getSender() {
+        return sender;
+    }
+
+    public Boolean flag() {
```
Comment (itsmarsss) [inline src/main/java/com/syncstream_dev/syncstream/model/message/UserMessage.java]: Functions/Methods that return a boolean type should be prefixed with `is` or `has`.

File: src/main/java/com/syncstream_dev/syncstream/model/ChatStream.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+import com.syncstream_dev.syncstream.model.message.UserMessage;
+import com.syncstream_dev.syncstream.model.user.User;
+
+public class ChatStream {
+    private HashMap<String, Message> messages;      // Message id to Message mapping
+    private ArrayList<User> participants;
+
+    public ChatStream() {
+        messages = new HashMap<String, Message>();
+        participants = new ArrayList<User>();
```
Comment (itsmarsss) [inline src/main/java/com/syncstream_dev/syncstream/model/ChatStream.java]: Java 17 can infer types: `new ArrayList<User>()` can just be `new ArrayList<>()`

We can keep it for the sake of readability.

File: src/main/java/com/syncstream_dev/syncstream/model/Genre.java [hunk 7]
```diff
@@ -0,0 +1,11 @@
+package com.syncstream_dev.syncstream.model;
+
+public enum Genre {
```
Comment (itsmarsss) [inline src/main/java/com/syncstream_dev/syncstream/model/Genre.java:3]: Reminder: "Dynamically" load genres into `Genre`. (We might have to use a `class` instead in the future)

File: src/main/java/com/syncstream_dev/syncstream/model/session/Session.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+public abstract class Session {
+    protected String sid;       // Session id
+    protected Genre genre;
+    protected ArrayList<User> participants;
+    protected ChatStream chatstream;
+    protected VideoInfo videoinfo;
+
+    // Create new session
+    public Session(String sid, Genre genre, ChatStream chatstream, VideoInfo videoinfo) {
```
Comment (itsmarsss) [inline src/main/java/com/syncstream_dev/syncstream/model/session/Session.java]: Variable convention: `camelCase`.

File: src/main/java/com/syncstream_dev/syncstream/model/user/User.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+        // Todo: Database integration
+        this.userInfo = new UserInfo(token, email, password, LocalDateTime.now(), LocalDateTime.now(), avatarUrl);
+        this.uid = uid;
+        this.username = username;
+        this.messagesSent = 0;
+        this.points = 0;
+    }
+
+    // Login User Constructor
+    public User(String username, String password) {
```
Comment (itsmarsss) [inline src/main/java/com/syncstream_dev/syncstream/model/user/User.java:24]: `password` is preferably a `char` or `byte` array.

Reminder for future: Encryption - hashing and salts.

File: src/main/java/com/syncstream_dev/syncstream/model/user/User.java [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+    public int getPoints() {
+        return points;
+    }
+
+    // Overrides
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) return true;
+        if (obj == null || getClass() != obj.getClass()) return false;
```
Comment (itsmarsss) [inline src/main/java/com/syncstream_dev/syncstream/model/user/User.java]: For future reference:
```java
public boolean equals(Object obj) {
     if (this == obj) return true; // If it's the same instance, they are equal
     if (!(obj instanceof User)) return false; // Check if the object is of type User
   
     User user = (User) obj; // Cast the object to User type
     return uid.equals(user.uid); // Check if uid fields are equal
}
```

File: src/main/java/com/syncstream_dev/syncstream/model/Genre.java (hunk 7 above)
Comment (itsmarsss) [inline src/main/java/com/syncstream_dev/syncstream/model/Genre.java:3]: Alphabetical order.

File: src/test/java/com/syncstream_dev/syncstream/datamodeltest/VideoInfoTest.java [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+    public void init() {
+        expectedUrl = "http://example.com";
+        expectedTitle = "Example Title";
+        expectedTimestamp = 123456789L;
+        videoInfo = new VideoInfo(expectedUrl, expectedTitle, expectedTimestamp);
+    }
+
+    @Test
+    public void testVideoInfoGetters() {
+
```
Comment (itsmarsss) [inline src/test/java/com/syncstream_dev/syncstream/datamodeltest/VideoInfoTest.java]: Spacing

**Review by itsmarsss — CHANGES_REQUESTED** [review]
View single comments, summary:
- [x] Avoid too much abbreviations (e.g. `mid` > `messageId`)
- [x] Java can infer ArrayList type (e.g. `new ArrayList<T>()` > `new ArrayList<>()`)
- [x] Boolean functions should be prefixed `is` or `has` (e.g. `cookies()` > `hasCookies()`)
- [x] Adhere to non-primitive types (e.g. `int` > `Integer`)
- [x] Variables should be `camelCased` (e.g. `chatstream` > `chatStream`)
- [x] Others... _(check specific comments)_

File: src/main/java/com/syncstream_dev/syncstream/model/user/User.java (hunk 9 above)
Comment (Xu-Gavin) [reply src/main/java/com/syncstream_dev/syncstream/model/user/User.java:24]: char array

File: src/main/java/com/syncstream_dev/syncstream/model/VideoInfo.java (hunk 1 above)
Comment (Xu-Gavin) [reply src/main/java/com/syncstream_dev/syncstream/model/VideoInfo.java]: 1a47af35ac9d322df9e9989e40623249abd61f78 addresses this

File: src/main/java/com/syncstream_dev/syncstream/model/message/Message.java (hunk 2 above)
Comment (Xu-Gavin) [reply src/main/java/com/syncstream_dev/syncstream/model/message/Message.java]: a747393f2eed0e5af27d61d62611a5176b43979a addresses this

File: src/main/java/com/syncstream_dev/syncstream/model/session/PrivateSession.java (hunk 3 above)
Comment (Xu-Gavin) [reply src/main/java/com/syncstream_dev/syncstream/model/session/PrivateSession.java:9]: We agreed to use primitive types, `Integer` -> `int` (which I already have) and `Long` -> `long`

Comment (Xu-Gavin) [reply src/main/java/com/syncstream_dev/syncstream/model/session/PrivateSession.java:9]: a2e382bdf3c2cdb06d1b89eef09f181dcc6e6ac9 addresses this

File: src/main/java/com/syncstream_dev/syncstream/model/user/User.java (hunk 4 above)
Comment (Xu-Gavin) [reply src/main/java/com/syncstream_dev/syncstream/model/user/User.java]: ddea308b2335cfae4542d79046a8495ecf95e8e1 addresses this

File: src/main/java/com/syncstream_dev/syncstream/model/message/UserMessage.java (hunk 5 above)
Comment (Xu-Gavin) [reply src/main/java/com/syncstream_dev/syncstream/model/message/UserMessage.java]: 7557948eb4b887d6619f0647c8e9d56e817770de addresses this

File: src/main/java/com/syncstream_dev/syncstream/model/ChatStream.java (hunk 6 above)
Comment (Xu-Gavin) [reply src/main/java/com/syncstream_dev/syncstream/model/ChatStream.java]: 612afa7f2a75cb4104ee3605adf069177b233d09 Addresses this

File: src/main/java/com/syncstream_dev/syncstream/model/session/Session.java (hunk 8 above)
Comment (Xu-Gavin) [reply src/main/java/com/syncstream_dev/syncstream/model/session/Session.java]: 561587fb17e4deed9055580433d3e6b8ba731c35 addresses this

File: src/main/java/com/syncstream_dev/syncstream/model/Genre.java (hunk 7 above)
Comment (Xu-Gavin) [reply src/main/java/com/syncstream_dev/syncstream/model/Genre.java:3]: Agreed, made a draft in the backlog for this.

File: src/main/java/com/syncstream_dev/syncstream/model/user/User.java (hunk 9 above)
Comment (Xu-Gavin) [reply src/main/java/com/syncstream_dev/syncstream/model/user/User.java:24]: 88cc85045b3b0d788200ab3e4557935c049ef00f addresses this

File: src/main/java/com/syncstream_dev/syncstream/model/user/User.java (hunk 10 above)
Comment (Xu-Gavin) [reply src/main/java/com/syncstream_dev/syncstream/model/user/User.java]: 753ba79180124ce26cca6b1553c442febd749ad7 addresses this

File: src/main/java/com/syncstream_dev/syncstream/model/Genre.java (hunk 7 above)
Comment (Xu-Gavin) [reply src/main/java/com/syncstream_dev/syncstream/model/Genre.java:3]: 09d9f9147b0cb8cd05efc45bd22ad5a9b7f50ba1 addresses this

File: src/test/java/com/syncstream_dev/syncstream/datamodeltest/VideoInfoTest.java (hunk 11 above)
Comment (Xu-Gavin) [reply src/test/java/com/syncstream_dev/syncstream/datamodeltest/VideoInfoTest.java]: Fixed

File: src/main/java/com/syncstream_dev/syncstream/model/ChatStream.java [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+        }
+
+        messages.put(msg.getMessageId(), msg);
+    }
+
+    public Message getMessage(String mid) {    // accepts message id (mid)
+        return messages.get(mid);
+    }
+
+    public Boolean userPresent(User user) {
```
Comment (itsmarsss) [inline src/main/java/com/syncstream_dev/syncstream/model/ChatStream.java]: Primitive type for `Boolean` -> `boolean`

**Review by itsmarsss — APPROVED** [review]
LGTM!!!!!!

**Pull request merged**